### PR TITLE
Fix #426 added feature: Play next

### LINKED ISSUE #426: Simple request: Play Song Next
It's currently possible to enqueue songs, but they always go to the very end of the playlist. If you're playing your entire library, that means you won't hear the song until the very end of the playlist. Please also add a button that queues the song to be played after the current song.

### PR DESCRIPTION
**Pull request by Yashraj254**
Fix #426 :  add a song next to the currently playing song in the queue...

Screenshot 2023-07-16 043451

https://github.com/SimpleMobileTools/Simple-Music-Player/assets/52420957/18caa448-7943-469a-94b7-373fde05131a

### CODE REVIEW

**Comment by tibbi** [comment]
@naveensingh please check if its ok

**Comment by naveensingh** [comment]
@Yashraj254 can you resolve the conflicts please

**Comment by naveensingh** [comment]
@Yashraj254 It crashes when I do **Play next** on **Albums**, **Artists** fragments. Do it in the background.
```
java.lang.IllegalStateException: Cannot access database on the main thread since it may potentially lock the UI for a long period of time.
    at androidx.room.RoomDatabase.assertNotMainThread(RoomDatabase.kt:444)
    at androidx.room.RoomDatabase.query(RoomDatabase.kt:485)
    at androidx.room.util.DBUtil.query(DBUtil.kt:75)
    at com.simplemobiletools.musicplayer.interfaces.AlbumsDao_Impl.getArtistAlbums(AlbumsDao_Impl.java:250)
    at com.simplemobiletools.musicplayer.adapters.ArtistsAdapter.getAllSelectedTracks(ArtistsAdapter.kt:112)
    at com.simplemobiletools.musicplayer.adapters.ArtistsAdapter.playNext(ArtistsAdapter.kt:102)
    at com.simplemobiletools.musicplayer.adapters.ArtistsAdapter.actionItemPressed(ArtistsAdapter.kt:65)
    at com.simplemobiletools.commons.adapters.MyRecyclerViewAdapter$1.onActionItemClicked(MyRecyclerViewAdapter.kt:58)
    ...
```

**Comment by naveensingh** [comment]
@tibbi do you think **Play next** should be visible when albums/artists are selected? We could add all the tracks in/by the selected album/artist to be played next but most players don't seem to allow this (**Play next** is only visible with tracks).

**Comment by Yashraj254** [comment]
should i remove this play next functionality for artists and albums and allow play next functionality for  single selected track only..

**Comment by naveensingh** [comment]
@Yashraj254 probably yes but let's wait for @tibbi's view...

**Comment by Yashraj254** [comment]
@naveensingh and currently Play Next functionality is available only if single track is selected, so is it fine ?

**Comment by naveensingh** [comment]
It is working fine for single tracks so far, the option is crashing at artists, albums fragments. I'm looking for code issues now...

**Comment by tibbi** [comment]
yeah, lets add it to tracks only

**Comment by naveensingh** [comment]
Edge cases:
- If I do **Play next** on the track _currently being played_, it moves to the top in the queue. We should keep the original order in that case (or we could disable the option for currently playing track altogether)

- Maybe we should synchronize access to `MusicService.mTracks`, it could throw `ConcurrentModificationException` in very rare situations.

Everything else looks good to me :+1:

**Comment by Yashraj254** [comment]
so, should i disable the Play Next option for currently playing track

**Comment by naveensingh** [comment]
yes, it doesn't work anyway. Note that `MusicService.mCurrentTrack` could be null in some cases

**Comment by tibbi** [comment]
so it is ready now?

**Comment by naveensingh** [comment]
Yes, it is ready. There's some minor style issues I commented above.

**Comment by Yashraj254** [comment]
Play next wont be visible if MusicService.mCurrentTrack is null

**Comment by naveensingh** [comment]
That's okay, otherwise, it'll probably crash something.

**Comment by naveensingh** [comment]
@tibbi this can be merged now, I'll fix any style issues in https://github.com/SimpleMobileTools/Simple-Music-Player/pull/559

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/adapters/AlbumsTracksAdapter.kt [hunk 1]
```diff
@@ -130,6 +137,22 @@ class AlbumsTracksAdapter(
...
 
+    private fun playNext() {
+        getSelectedTracks().firstOrNull()?.let { selectedTrack ->
+            activity.playNextInQueue(selectedTrack) {
+                finishActMode()
+            }
+        }
+    }
+
+    private fun isATrack(): Boolean {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/musicplayer/adapters/AlbumsTracksAdapter.kt:148]: rename this please, it doesnt make much sense

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/adapters/AlbumsTracksAdapter.kt [hunk 2]
```diff
@@ -130,6 +137,22 @@ class AlbumsTracksAdapter(
...
+        getSelectedItems().forEach {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/musicplayer/adapters/AlbumsTracksAdapter.kt:149]: you can use any() or none() for these

**Comment by tibbi** [comment]
fixed it myself to speed it up, thanks

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/adapters/AlbumsTracksAdapter.kt (hunk 2 above)
Comment (Yashraj254) [reply app/src/main/kotlin/com/simplemobiletools/musicplayer/adapters/AlbumsTracksAdapter.kt]: Understood. My mistake. Apologies for any unnecessary complications.

**Pull request merged**